## 1. Counting missing values

<h3 id="info"><code>info</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>description</code></td>
<td><code>varchar</code></td>
<td>Description of the product</td>
</tr>
</tbody>
</table>
<h3 id="finance"><code>finance</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>listing_price</code></td>
<td><code>float</code></td>
<td>Listing price for product</td>
</tr>
<tr>
<td><code>sale_price</code></td>
<td><code>float</code></td>
<td>Price of the product when on sale</td>
</tr>
<tr>
<td><code>discount</code></td>
<td><code>float</code></td>
<td>Discount, as a decimal, applied to the sale price</td>
</tr>
<tr>
<td><code>revenue</code></td>
<td><code>float</code></td>
<td>Amount of revenue generated by each product, in US dollars</td>
</tr>
</tbody>
</table>
<h3 id="reviews"><code>reviews</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>rating</code></td>
<td><code>float</code></td>
<td>Product rating, scored from <code>1.0</code> to <code>5.0</code></td>
</tr>
<tr>
<td><code>reviews</code></td>
<td><code>float</code></td>
<td>Number of reviews for the product</td>
</tr>
</tbody>
</table>
<h3 id="traffic"><code>traffic</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>last_visited</code></td>
<td><code>timestamp</code></td>
<td>Date and time the product was last viewed on the website</td>
</tr>
</tbody>
</table>
<h3 id="brands"><code>brands</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>brand</code></td>
<td><code>varchar</code></td>
<td>Brand of the product</td>
</tr>
</tbody>
</table>

In [171]:
%%sql
postgresql:///sports

select count(*) as total_rows,
COUNT(i.description) AS count_description, 
    COUNT(f.listing_price) AS count_listing_price, 
    COUNT(t.last_visited) AS count_last_visited 
from info as i
join finance as f
using(product_id)
join traffic as t
using(product_id)


1 rows affected.


[(3179, 3117, 3120, 2928)]

## 2. Nike vs Adidas pricing

In [ ]:
%%sql

select brand, listing_price ::int, count(f.product_id)
from brands as b
inner join finance as f
using(product_id)
where listing_price > 0
group by brand,  listing_price
order by 2 desc

        

 * postgresql:///sports
77 rows affected.


[('Adidas', 300, 2),
 ('Adidas', 280, 4),
 ('Adidas', 240, 5),
 ('Adidas', 230, 8),
 ('Adidas', 220, 11),
 ('Nike', 200, 1),
 ('Adidas', 200, 8),
 ('Nike', 190, 2),
 ('Adidas', 190, 7),
 ('Nike', 180, 4),
 ('Adidas', 180, 34),
 ('Nike', 170, 14),
 ('Adidas', 170, 27),
 ('Nike', 160, 31),
 ('Adidas', 160, 28),
 ('Adidas', 150, 41),
 ('Nike', 150, 6),
 ('Adidas', 140, 36),
 ('Nike', 140, 12),
 ('Adidas', 130, 96),
 ('Nike', 130, 12),
 ('Adidas', 120, 115),
 ('Nike', 120, 16),
 ('Nike', 110, 17),
 ('Adidas', 110, 91),
 ('Nike', 100, 14),
 ('Adidas', 100, 72),
 ('Adidas', 96, 2),
 ('Nike', 95, 1),
 ('Nike', 90, 13),
 ('Adidas', 90, 89),
 ('Adidas', 86, 7),
 ('Nike', 85, 5),
 ('Adidas', 85, 1),
 ('Adidas', 80, 322),
 ('Nike', 80, 16),
 ('Nike', 79, 1),
 ('Adidas', 76, 149),
 ('Adidas', 75, 1),
 ('Nike', 75, 7),
 ('Adidas', 70, 87),
 ('Nike', 70, 4),
 ('Adidas', 66, 102),
 ('Nike', 65, 1),
 ('Adidas', 63, 1),
 ('Nike', 60, 2),
 ('Adidas', 60, 211),
 ('Adidas', 56, 174),
 ('Adidas', 55, 2),
 ('Adidas', 53, 43),
 ('Nike', 50, 5),
 ('Adidas', 50, 183),
 ('Adidas', 48, 42),
 ('Nike', 48, 1),
 ('Adidas', 46, 163),
 ('Nike', 45, 3),
 ('Adidas', 45, 1),
 ('Adidas', 43, 51),
 ('Adidas', 40, 81),
 ('Nike', 40, 1),
 ('Adidas', 38, 24),
 ('Adidas', 36, 25),
 ('Adidas', 33, 24),
 ('Adidas', 30, 37),
 ('Nike', 30, 2),
 ('Adidas', 28, 38),
 ('Adidas', 27, 18),
 ('Adidas', 25, 28),
 ('Adidas', 23, 1),
 ('Adidas', 20, 8),
 ('Adidas', 18, 4),
 ('Adidas', 16, 4),
 ('Adidas', 15, 27),
 ('Adidas', 13, 27),
 ('Adidas', 12, 1),
 ('Adidas', 10, 11),
 ('Adidas', 9, 1)]

## 3. Labeling price ranges

In [175]:
%%sql


select brand, count(f.product_id), sum(revenue) as total_revenue,
case when listing_price <42 then 'Budget'
when listing_price <74 then 'Average'
when listing_price < 129 then 'Expensive'
else 'Elite' End as price_category
from brands as b
inner join finance as f
using(product_id)
where brand is not null 
group by brand, price_category
order by total_revenue desc


 * postgresql:///sports
8 rows affected.


[('Adidas', 849, 4626980.069999999, 'Expensive'),
 ('Adidas', 1060, 3233661.060000001, 'Average'),
 ('Adidas', 307, 3014316.8299999987, 'Elite'),
 ('Adidas', 359, 651661.1200000002, 'Budget'),
 ('Nike', 357, 595341.0199999992, 'Budget'),
 ('Nike', 82, 128475.59000000003, 'Elite'),
 ('Nike', 90, 71843.15000000004, 'Expensive'),
 ('Nike', 16, 6623.5, 'Average')]

## 4. Average discount by brand

<p>Note we have been looking at <code>listing_price</code> so far. The <code>listing_price</code> may not be the price that the product is ultimately sold for.  let's take a look at the <code>discount</code>, which is the percent reduction in the <code>listing_price</code> when the product is actually sold. We would like to know whether there is a difference in the amount of <code>discount</code> offered between brands, as this could be influencing <code>revenue</code>.</p>

In [177]:
%%sql
    
SELECT b.brand, AVG(f.discount) * 100 AS average_discount
FROM brands AS b
INNER JOIN finance AS f 
    ON b.product_id = f.product_id
GROUP BY b.brand
HAVING b.brand IS NOT NULL
ORDER BY average_discount;

 * postgresql:///sports
2 rows affected.


[('Nike', 0.0), ('Adidas', 33.452427184465606)]

## 5. Correlation between revenue and reviews

In [179]:
%%sql

select corr(reviews, revenue) as review_revenue_corr
from reviews
inner join finance
using(product_id)

 * postgresql:///sports
1 rows affected.


[(0.6518512283481301,)]

## 6. Ratings and reviews by product description length


In [181]:
%%sql


-- Filter for products without missing values for description, and sort results by description_length


SELECT TRUNC(LENGTH(i.description), -2) AS description_length,
    ROUND(AVG(r.rating::numeric), 2) AS average_rating
FROM info AS i
INNER JOIN reviews AS r 
    ON i.product_id = r.product_id
WHERE i.description IS NOT NULL
GROUP BY description_length
ORDER BY description_length;



 * postgresql:///sports
7 rows affected.


[(Decimal('0'), Decimal('1.87')),
 (Decimal('100'), Decimal('3.21')),
 (Decimal('200'), Decimal('3.27')),
 (Decimal('300'), Decimal('3.29')),
 (Decimal('400'), Decimal('3.32')),
 (Decimal('500'), Decimal('3.12')),
 (Decimal('600'), Decimal('3.65'))]

## 7. Reviews by month and brand

<p>Let's take a look at the volume of <code>reviews</code> by month to see if there are any trends or gaps we can look to exploit.</p>

In [183]:
%%sql


select brand, date_part('month', t.last_visited ) as month, count(r.product_id) as num_reviews
from brands
join reviews as r
using(product_id)
join traffic as t
using(product_id)
group by brand, month
having brand is not null and date_part('month', t.last_visited )is not null 
order by brand, month

 * postgresql:///sports
24 rows affected.


[('Adidas', 1.0, 253),
 ('Adidas', 2.0, 272),
 ('Adidas', 3.0, 269),
 ('Adidas', 4.0, 180),
 ('Adidas', 5.0, 172),
 ('Adidas', 6.0, 159),
 ('Adidas', 7.0, 170),
 ('Adidas', 8.0, 189),
 ('Adidas', 9.0, 181),
 ('Adidas', 10.0, 192),
 ('Adidas', 11.0, 150),
 ('Adidas', 12.0, 190),
 ('Nike', 1.0, 52),
 ('Nike', 2.0, 52),
 ('Nike', 3.0, 55),
 ('Nike', 4.0, 42),
 ('Nike', 5.0, 41),
 ('Nike', 6.0, 43),
 ('Nike', 7.0, 37),
 ('Nike', 8.0, 29),
 ('Nike', 9.0, 28),
 ('Nike', 10.0, 47),
 ('Nike', 11.0, 38),
 ('Nike', 12.0, 35)]

## 8. Footwear product performance

<p>So far, we have been primarily analyzing Adidas vs Nike products. Now, let's switch our attention to the type of products being sold. As there are no labels for product type, we will create a Common Table Expression (CTE) that filters <code>description</code> for keywords, then use the results to find out how much of the company's stock consists of footwear products and the median <code>revenue</code> generated by these items.</p>

In [185]:
%%sql


with footwear as (
SELECT i.description, f.revenue
    FROM info AS i
    INNER JOIN finance AS f 
        ON i.product_id = f.product_id
    WHERE i.description ILIKE '%shoe%'
        OR i.description ILIKE '%trainer%'
        OR i.description ILIKE '%foot%'
        AND i.description IS NOT NULL
)

SELECT COUNT(*) AS num_footwear_products, 
    percentile_disc(0.5) WITHIN GROUP (ORDER BY revenue) AS median_footwear_revenue
FROM footwear;

 * postgresql:///sports
1 rows affected.


[(2700, 3118.36)]

## 9. Clothing product performance
<p> we found there are 3,117 products without missing values for <code>description</code>. Of those, 2,700 are footwear products, which accounts for around 85% of the company's stock. They also generate a median revenue of over $3000 dollars!</p>
<p>So, for our final task, let's examine how this differs to clothing products. We will re-use <code>footwear</code>, adding a filter afterward to count the number of products and <code>median_revenue</code> of products that are not in <code>footwear</code>.</p>

In [187]:
%%sql

-- Copy the footwear CTE from the previous task
-- Calculate the number of products in info and median revenue from finance
-- Inner join info with finance on product_id
-- Filter the selection for products with description not in footwear

with footwear as (
SELECT i.description, f.revenue
    FROM info AS i
    INNER JOIN finance AS f 
        ON i.product_id = f.product_id
    WHERE i.description ILIKE '%shoe%'
        OR i.description ILIKE '%trainer%'
        OR i.description ILIKE '%foot%'
        AND i.description IS NOT NULL
)

SELECT COUNT(i.*) AS num_clothing_products, 
    percentile_disc(0.5) WITHIN GROUP (ORDER BY f.revenue) AS median_clothing_revenue
FROM info AS i
INNER JOIN finance AS f on i.product_id = f.product_id
WHERE i.description NOT IN (SELECT description FROM footwear);



 * postgresql:///sports
1 rows affected.


[(417, 503.82)]